src/python/pants/util/objects.py

```diff
@@ -52,6 +58,394 @@ def __getnewargs__(self):
+    if cls.LOWER_CASE_TYPE_NAME.match(type_name):
+      # double underscore here ensures no clash with camel-case type names
+      return 'primitive__{}'.format(type_name)
+
+    if cls.CAMEL_CASE_TYPE_NAME.match(type_name):
+      split_by_camel_downcased = []
+      for m in cls.CAMEL_CASE_SPLIT_PATTERN.finditer(type_name):
+        camel_group = m.group(0)
+        downcased = camel_group.lower()
+        split_by_camel_downcased.append(downcased)
+      return '_'.join(split_by_camel_downcased)
+
+    raise cls.FieldTypeNameError(
+      "Type name {!r} must be camel-cased with an initial capital, "
+      "or all lowercase. Only ASCII alphabetical characters are allowed."
+      .format(type_name))
+
+  def __init__(self, single_type, field_name):
+    if not isinstance(single_type, type):
+      raise self.FieldTypeConstructionError(
+        "single_type is not a type: was {!r} (type {!r})."
+        .format(single_type, type(single_type).__name__))
+    if not isinstance(field_name, str):
+      raise self.FieldTypeConstructionError(
+        "field_name is not a str: was {!r} (type {!r})."
+        .format(field_name, type(field_name).__name__))
+
+    super(FieldType, self).__init__(single_type)
+
+    self._field_name = field_name
+
+  @property
+  def field_name(self):
+    return self._field_name
+
+  @property
+  def field_type(self):
+    return self.types[0]
+
+  def validate_satisfies_field(self, obj):
+    """Return `obj` if it satisfies this type constraint, or raise.
+
+    Use this method over `self.validate_satisfied_by()` to provide a more
+    specific error message for the FieldType class.
+
+    :raises: `TypeConstraintError` if the given object does not satisfy this
+    type constraint.
+    """
+    if self.satisfied_by(obj):
+      return obj
+
+    raise TypeConstraintError(
+      "value {!r} (with type {!r}) must be an instance of type {!r}."
+      .format(obj, type(obj).__name__, self.field_type.__name__))
+
+  def __repr__(self):
+    fmt_str = 'FieldType({field_type}, {field_name!r})'
+    return fmt_str.format(field_type=self.field_type.__name__,
+                          field_name=self.field_name)
+
+  @classmethod
+  def create_from_type(cls, type_obj):
+    """Generate a FieldType with a predictable name string from the given type.
+
+    The field name will be a deterministic and unique string generated from the
+    type `type_obj`. `type_obj` must have a camel-cased name (e.g. MyType) or
+    all-lowercased name (e.g. int).
+    """
+    if not isinstance(type_obj, type):
+      raise cls.FieldTypeConstructionError(
+        "type_obj is not a type: was {!r} (type {!r})"
+        .format(type_obj, type(type_obj).__name__))
+    transformed_type_name = cls._transform_type_field_name(type_obj.__name__)
+    return cls(type_obj, str(transformed_type_name))
+
+
+def typed_datatype(type_name, field_decls):
+  """A wrapper over namedtuple which accepts a dict of field names and types.
```

Comment:
This no longer takes a dict.

Reply:
I moved the contents of this docstring to the docstring of the `datatype()` function and added a `FIXME:` to edit it later (and "later" should be right now) in 2d1489e.

Reply:
After rebasing, the sha is 6420f5b. I removed the FIXME in 378a75e.